src/assets/AccountTrackerController.ts

```diff
@@ -146,14 +146,37 @@ export class AccountTrackerController extends BaseController<
    */
   refresh = async () => {
     this.syncAccounts();
-    const { accounts } = this.state;
+    const accounts = Object.assign({}, this.state.accounts);
+    for (const address in accounts) {
+      await safelyExecuteWithTimeout(async () => {
+        const balance = await query(this.ethQuery, 'getBalance', [address]);
+        accounts[address] = { balance: BNToHex(balance) };
+      });
+    }
+    this.update({ accounts: { ...accounts } });
+  };
+
+  /**
+   * Sync accounts balances with some additional addresses.
+   *
+   * @param addresses - the additional addresses, may be hardware wallet addresses.
+   * @returns accounts - current state accounts
+   */
+  syncWithAddresses = async (addresses: string[]) => {
+    this.syncAccounts();
+    const accounts = Object.assign({}, this.state.accounts);
+    addresses.forEach((address) => {
+      accounts[address] = { balance: '0x0' };
+    });
+
     for (const address in accounts) {
       await safelyExecuteWithTimeout(async () => {
         const balance = await query(this.ethQuery, 'getBalance', [address]);
         accounts[address] = { balance: BNToHex(balance) };
-        this.update({ accounts: { ...accounts } });
       });
     }
```

Comment:
Is there a reason why you're setting all account balances to 0 and then requesting the balance of each account? Additionally you should be able to shorten this without needing to clone anything:

``` suggestion
    const accounts = await Promise.all(
      accounts.map((address) => {
        return safelyExecuteWithTimeout(async () => {
          const balance = await query(this.ethQuery, 'getBalance', [address]);
          return [address, balance];
        })
      })
    ).reduce((obj, [address, balance]) => {
      return { ...obj: [address]: balance }
    }, {});
```

Reply:
Hi @mcmire , this method is for syncing accounts balance with some additional addresses. So I update the cloned accounts with the additional addresses firstly. 
But, actually the state update is not necessary here I think, I will modify this method to a pure function just return the additional addresses with synced balance.